### PR TITLE
Esb-152

### PR DESCRIPTION
**Pull request by solglass**
User repository positive tests.
GetUsers() is now impossible to test correctly (if necessary) (not an IUserRepository interface method)
GetPassedStudentsAttempt_SelectByGroupId 1) doesn't seem to belong in repo test (ex-Homework 
attempt)
2) needs to be tested in appropriate report service

### CODE REVIEW

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            Assert.Greater(addedEntityId, 0);
+            dto.Roles = new List<Role> { Role.Admin };
+            _addedUserRoleIds.Add((addedEntityId, (int)Role.Admin));
+
+            //When
+            _addedUserDtoIds.Add(addedEntityId);
+
+            // Then
+            dto.Id = addedEntityId;
+            var actual = _repository.GetUserById(addedEntityId);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:89]: This string is "when"

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            Assert.AreEqual(dto, actual);
+        }
+
+
+        [TestCase(1)]
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:94]: Why dont you create two testcases with deleting and with recovering?

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            var dto = (UserDto)UserMockGetter.GetUserDtoMock(mockId).Clone();
+            var addedEntityId = _repository.AddUser(dto);
+            _repository.AddRoleToUser(addedEntityId, 1);
+            dto.Roles = new List<Role> { Role.Admin };
+            _addedUserDtoIds.Add(addedEntityId);
+            dto.Id = addedEntityId;
+            _addedUserRoleIds.Add((addedEntityId, (int)Role.Admin));
+
+            //When
+            _repository.DeleteOrRecoverUser(addedEntityId, true);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:107]: Assertion that amount affected rows is 1?

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+            // Then
+            dto.IsDeleted = true;
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(dto, actual);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:112]: Dto to expected

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+                IsDeleted = dto.IsDeleted,
+                Roles = dto.Roles
+        }; 
+
+            //When
+            _repository.UpdateUser(dto);
+
+            // Then
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(dto, actual);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:148]: Dto to expected

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            _repository.AddRoleToUser(addedEntityId, 1);
+            _repository.AddRoleToUser(addedEntityId, 2);
+            _repository.AddRoleToUser(addedEntityId, 3);
+
+            _addedUserRoleIds.Add((addedEntityId, 1));
+            _addedUserRoleIds.Add((addedEntityId, 2));
+            _addedUserRoleIds.Add((addedEntityId, 3));
+
+            //When
+            _repository.DeleteRoleToUser(addedEntityId, 2);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:171]: Delete role from user))

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+            // Then
+            dto.Id = addedEntityId;
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:174]: Why this is in "then" part?

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+            dto.Roles = new List<Role> { Role.Admin,Role.Teacher };
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(dto, actual);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:177]: Dto to expected and check userDto equals method

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+            _addedUserRoleIds.Add((addedEntityId, 1));
+            string newPassword = "fskljjsdljf";
+
+            //When
+            _repository.ChangeUserPassword(addedEntityId, dto.Password, newPassword);
+
+            // Then
+            dto.Password = newPassword;
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(dto, actual);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:225]: dto to expected

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+            _addedUserDtoIds.Add(addedEntityId);
+            _addedUserRoleIds.Add((addedEntityId, 1));
+            string newPassword = "fskljjsdljf";
+
+            //When
+            _repository.ChangeUserPassword(addedEntityId, dto.Password, newPassword);
+
+            // Then
+            dto.Password = newPassword;
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:223]: This should be in when-section

**Comment by solglass** [comment]
Improved after review with additions and fixes to the mentioned methods

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            dto.Roles = new List<Role> { Role.Admin };
+            _addedUserDtoIds.Add(addedEntityId);
+            _addedUserRoleIds.Add((addedEntityId, (int)Role.Admin));
+            dto.Id = addedEntityId;
+
+            //When
+            var actual = _repository.GetUserById(addedEntityId);
+
+            // Then
+            Assert.AreEqual(dto, actual);
```
[the remark this answers is not in the document]
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: renamed

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+        }
+
+        [TestCase(new int[] { 1, 2, 3 })]
+        public void UserGetUsersPositiveTest(int[] mockIds)
+        {
+            //Given
+            var dtoList = new List<UserDto>();
```
[the remark this answers is not in the document]
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: Fixed in setup. Test base must be empty no matter what.

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+                _addedUserRoleIds.Add((addedEntityId, (int)Role.Admin));
+                dto.Id = addedEntityId;
+                dtoList.Add(dto);
+            }
+
+            //When
+            var actual = _repository.GetUsers();
+
+            // Then
+            Assert.AreEqual(dtoList, actual);
```
[the remark this answers is not in the document]
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: fixed

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 1 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs:96]: changed to "(When, Then)"

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 2 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs:102]: Expanded

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 6 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: renamed

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 7 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: changed to "(When, Then)"

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 8 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: renamed

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 9 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: renamed

File: EducationSystem/EducationSystem.Test/UserTests.cs (hunk 10 above)
Comment (solglass) [reply EducationSystem/EducationSystem.Test/UserTests.cs]: changed to "(When, Then)"

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+using EducationSystem.Core.Enums;
+using System.Linq;
+
+namespace EducationSystem.Data.Tests
+{
+    public class UserTests : BaseTest
+    {
+        private List<int> _addedUserDtoIds;
+        private List<(int, int)> _addedUserRoleIds;
+        private UserRepository _repository;
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:17]: Use interface

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+
+        [OneTimeSetUp]
+        public void UserRepositoryTestsSetup()
+        {
+            _addedUserDtoIds = new List<int>();
+            _repository = new UserRepository(_options);
+            _addedUserRoleIds = new List<(int, int)>();
+            var junk = _repository.GetUsers();
+            if (junk.Count > 0)
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:26]: Haha nice

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        //  entity tests are below:
+
+        [TestCase(1)]
+        public void UserGetUserByIdPositiveTest(int mockId)
+        {
+            //Given
+            var expected = (UserDto)UserMockGetter.GetUserDtoMock(mockId).Clone();
+            var addedEntityId = _repository.AddUser(expected);
+            _repository.AddRoleToUser(addedEntityId, 1);
+            expected.Roles = new List<Role> { Role.Admin };
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:49]: And you could hide the initialization in the GetMock method, now it is ok, but if you are planning to initialize list of roles in every test it is better to write it once

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+                expected.Add(dto);
+            }
+
+            //When, Then
+            var actual = _repository.GetUsers();
+            CollectionAssert.AreEqual(expected, actual);
+
+            _addedUserDtoIdsLocal.ForEach(id =>
+            {
+                CollectionAssert.AreEqual(expected.First(item => item.Id == id).Roles, actual.First(item => item.Id == id).Roles);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:87]: Cant understand why first?

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+            _addedUserDtoIds.Add(addedEntityId);
+            expected.Id = addedEntityId;
+            _addedUserRoleIds.Add((addedEntityId, (int)Role.Admin));
+
+            //When, Then
+            _repository.DeleteOrRecoverUser(addedEntityId, true);
+            expected.IsDeleted = true;
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(expected, actual);
+            CollectionAssert.AreEqual(expected.Roles, actual.Roles);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:133]: Cannot understand if you just check the workability of select by id or that roles didn't change during updating isDeleting

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+                IsDeleted = expected.IsDeleted,
+                Roles = expected.Roles
+        };
+
+            //When, Then
+            _repository.UpdateUser(expected);
+
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(expected, actual);
+            CollectionAssert.AreEqual(expected.Roles, actual.Roles);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:169]: Same

File: EducationSystem/EducationSystem.Test/UserTests.cs [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+            _addedUserRoleIds.Add((addedEntityId, 1));
+            string newPassword = "fskljjsdljf";
+
+            //When, Then
+            _repository.ChangeUserPassword(addedEntityId, expected.Password, newPassword);
+
+            expected.Password = newPassword;
+            var actual = _repository.GetUserById(addedEntityId);
+            Assert.AreEqual(expected, actual);
+            CollectionAssert.AreEqual(expected.Roles, actual.Roles);
```
Comment (Zhanna-ZhuZhanna) [inline EducationSystem/EducationSystem.Test/UserTests.cs:246]: Same

**Pull request merged**